### ISSUE TITLE
[FEATURE REQUEST] Integration with org-roam and Zotero

### ISSUE THREAD

**Issue by wakatara** [org-noter/org-noter#2]
Perhaps the biggest change since my workflow when I was last using org-noter and before it snew update, is the use of org-roam to take notes.

It would be amazing if org-note oculd be org-roam aware so that when you do start to take notes on a pdf they could be asked for an org-roam capture tempalte and the pdf linked to the org-roam note. Would be rather epic. (and I know many people are using org-roam for precisely this type of knowledge management these days.).

**Comment by petermao** [comment]
I don't use org-roam (yet?) so I don't know what this would entail.  what is a "capture template"?

From the org-roam user point of view, if they are already using it to link notes to pdfs, what does org-noter bring to the table for them?

I'd be happy to review a PR or help you get started with this feature, but I'm not the one to write it!

**Comment by petermao** [comment]
@bdarcus mentioned on r/emacs that [org-roam-bibtex](https://github.com/org-roam/org-roam-bibtex) has some noter integration: https://github.com/org-roam/org-roam-bibtex/blob/main/doc/orb-manual.org#org-noter-integration-special-treatment-of-the-file-keyword

**Comment by wakatara** [comment]
Apologies @petermao, I'm realizing I expressed this poorly. What I would *like* to be able to do is this following and have org-noter work as part of the note taking flow. I guess I'm just trying to figure out how to have that suported.

1. Open a pdf from org-roam in Zotero via its ability to create insert-a-new link
2. Use org-noter for the Zotero stored pdf to take notes within the org-roam file
3. When I go back to my notes be able to get back to the typical org-noter - here's the ref in the pdf feature

To answer your questions, org-roam is pretty nifty but really just something that sits on org-mode to provide bi-directional linking (and collection) of links into a document. Super handy and built on the idea of zettelkasten. If you've ever seen roam research, obsidian, or logseq, it works similarly. 

A capture template allows you (when creating a new link  like say to) to open org-capture with a specific set of templated fields (as well as filename/directory etc - in this case I am hoping with the citekey) and have a template filled in. Kinda like traditional org-capture templates you're probably familiar with but with more features. 

(and apologies for expressing the feature request poorly at first.).

**Comment by chatziiola** [comment]
I'm not entirely sure I understand @wakatara 's workflow, and I'm really sorry if this is not the place to post an opinion.

Even though this is not ideal, and I had tried to find of a better solution, I do not think that we need `org-roam`'s capture templates. 


The problem, from my point of view, stems from ~org-noter~ 's inability to work with top level properties (such as the `:PROPERTIES:` drawer before the file `#+TITLE`). If we were to add support so that one text note from org noter was of this form:

```org
:PROPERTIES:
:NOTER_DOCUMENT: /path/to/document
:END:
#+TITLE: ...
```

And not of the current form:
```org
#+TITLE:  ...
...
* Some top level header
:PROPERTIES:
:NOTER_DOCUMENT: /path/to/document
:END:
```

We could use the filename in a *standardized* form as an `org-roam` id. Keep in mind that IDs are used like:
```org
:PROPERTIES:
:ID:       16EA43E1-7DE2-4F9C-9731-BBB845854018
:END:
#+title: ...
```
And `org-roam` 's db can be updated by ~org-roam-db-update` if I'm not mistaken.

(This way if someone plays around with packages that use slightly different *properties* not a lot of code should be changed/created.)

I have not written about citekeys here, because I'm afraid that I have only a shallow understanding of their workings. In my **sketchy** so far set up, practically doing what I described above, I use this function, and then I create a top level heading with  the necessary property`:NOTER_DOCUMENT: /path/to/document`:

```elisp
    (defun chatziiol/org-roam-node-from-cite (keys-entries)
    "Inspired by https://jethrokuan.github.io/org-roam-guide/.
    Basically a working re-implementation of that function"
	(interactive (list (citar-select-ref)))
	(let ((title (citar-format--entry "${author editor}: ${title}" keys-entries nil)))
	(org-roam-capture- :templates
			    '(("g" "general" plain "%?" :if-new
				(file+head "braindump/${citekey}.org"
                                        ":PROPERTIES:\n:ROAM_REFS: cite:@${citekey}\n:END:\n#+title: ${title}\n")
				:immediate-finish t
				:unnarrowed t))
			    :info (list :citekey keys-entries)
			    :node (org-roam-node-create :title title)
			    :props '(:finalize find-file)))))
```

**Comment by bdarcus** [comment]
Am not really following all the details, but on these:

> The problem, from my point of view, stems from org-noter 's inability to work with top level properties.

That does seem a problem ;-)

> I have not written about citekeys here, because I'm afraid that I have only a shallow understanding of their workings.

It's just a local key for a reference.

In org-roam, adding one as a roam_ref property allows org-roam to identify it was a bibliographic note, and so associate it with that reference.

Those references are actually stored in the org-roam db. So in citar-org-roam, for example, we query the DB for notes of that type, and to associate references in citar with those notes.

HTH.

**Comment by petermao** [comment]
> And not of the current form:
> 
> ```
> #+TITLE:  ...
> ...
> * Some top level header
> :PROPERTIES:
> :NOTER_DOCUMENT: /path/to/document
> :END:
> ```
`Some top level header` you have there is `<document>` by default in org-noter.  This allows the notes for multiple documents to sit in a single org files ("Notes.org" by default), a feature that was built in by @weirdNox and which I rely on heavily in my own workflow.

**Comment by petermao** [comment]
> Apologies @petermao, I'm realizing I expressed this poorly. What I would _like_ to be able to do is this following and have org-noter work as part of the note taking flow. I guess I'm just trying to figure out how to have that suported.
> 
>     1. Open a pdf from org-roam in Zotero via its ability to create insert-a-new link
> 
>     2. Use org-noter for the Zotero stored pdf to take notes within the org-roam file
> 
>     3. When I go back to my notes be able to get back to the typical org-noter - here's the ref in the pdf feature
@wakatara, thanks for the clarification, but this feature is unlikely to come from me, but we would definitely review it in good faith if some enterprising user were to take it up.

**Comment by dmitrym0** [comment]
We could certainly integrate org-noter and org-roam (or standard org-templates) but I'm not sure how that would be useful. @wakatara what kinds of templates do you envision using with noter?

top level headings: I was initially confused by how org-noter keeps all notes in a single file but I've grown used to it. Architecturally I don't believe there is anything preventing from supporting top level `:NOTER_DOCUMENT` property. It would be more convenient for org-roam. In fact that's how the rest of my knowledge. based organized -- hundreds of tiny files. Meanwhile, it's possible to create IDs for headings inside the notes file, and org-roam will pick that up for linking.

If anyone wants to have a go at it 👍

**Comment by petermao** [comment]
@wakatara, did you find a solution?  If not, it's fine to leave the issue open in case some brilliant Alyssa P. Hacker comes by and solve it for everyone.
